### DISCUSSION TITLE
What is the best practice to loading content in the background?

### DISCUSSION THREAD

**Discussion by Im-Oab** [17cupsofcoffee/tetra#228] [Q&A] [answered]
Hi, I just curious about What is the best practice to loading content in the background? 

Currently, I load 10 files per update() and if accumulated time more than 1 second. It will not continue loading files until accumulated time back to normal. This allows my scene to have a loading animation.

**Comment by 17cupsofcoffee** [comment, marked as answer]
I think the ideal solution is to have assets load on a seperate thread - you can then use something like a `channel` or an atomic to communicate back to the main thread what the progress is.

The important thing to consider is that `Context` can't be accessed unless you're on the main thread - so for types like `Texture` which need to access `Context`, you'll need to pass the loaded file back across to the main thread to create your final asset (e.g. via `Texture::from_file_data`).

I'll try to write up an example of what this could look like, it's something I've been meaning to do for a while!

**Reply by Im-Oab** [reply]
Last time, I attempted for loading in the background. I stuck with Texture and Context. :')

Also, I appreciate your examples. My game's codes come from study your examples.

**Reply by 17cupsofcoffee** [reply]
Here's what I was able to come up with:

https://gist.github.com/17cupsofcoffee/bb57dc1ebfe6507a262513117d0b1901

As mentioned in the comments, this isn't the only (or even necessarily the *right*) way to do this, but it should hopefully give you some ideas :smile: I've not worked on any projects big enough to really need a proper loading screen yet, so this was a good learning experience for me.
